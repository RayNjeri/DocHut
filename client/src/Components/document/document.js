import React from 'react';
import PropTypes from 'prop-types';

export default class Documents extends React.Component{
  constructor(props, context){
    super(props, context);
    this.state={
      document:{content:null}
    };
  }

  onContentChange(event){
    const content = this.state.content;
    document.content = event.target.value;
    this.setState({content: content});
  }

  render(){
    return(
      <div>
        <h1> Documents </h1>
        <h2> Create documents </h2>
        <input type ='text' Onchange={this.onContentChange} value={this.state.document.content}/>
        <input type='submit' value= 'Save' onClick={this.onClickSave}/>
      </div>

        );
  }
}
