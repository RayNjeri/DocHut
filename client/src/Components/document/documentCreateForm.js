import React from 'react';
import PropTypes from 'prop-types';
import TextField from 'material-ui/TextField';
import SelectField from 'material-ui/SelectField';
import MenuItem from 'material-ui/MenuItem';
import DocumentView from './documentsView';

const CreateDocument = props => (
  <div>
    <TextField
      name="title"
      type="text"
      onChange={props.onTitleChange}
      defaultValue={props.document.title}
      fullWidth
    /> <br /><br />

    <TextField
      className="content"
      name="content"
      type="text"
      onChange={props.onContentChange}
      defaultValue={props.document.content}
      fullWidth
      multiLine
      rows={2}
      rowsMax={10}
    /><br /> <br />

    <SelectField
      floatingLabelText="Access"
      value={props.document.access}
      onChange={props.onSetAccess}
    >
      <MenuItem value={'public'} primaryText="Public" />
      <MenuItem value={'private'} primaryText="Private" />

    </SelectField>
  </div>
);

export default CreateDocument;
