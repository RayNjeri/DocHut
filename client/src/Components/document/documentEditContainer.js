import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import FlatButton from 'material-ui/FlatButton';
import Dialog from 'material-ui/Dialog';
import documentActions from '../../actions/documentsAction';
import { updateDocument, deleteDocument, listDocuments } from '../../actions/documentsAction';
import EditDocument from './documentForm';

class EditDocumentContainer extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            open: true,
            document: {
                id: props.document.id,
                title: props.document.title,
                content: props.document.content,
                access: props.document.access
            }
        };
        this.state.open = true;
        this.onSubmit = this.onSubmit.bind(this);
        this.updateDocument = this.updateDoc.bind(this);
        this.handleOpen = this.handleOpen.bind(this);
        this.handleClose = this.handleClose.bind(this);
    }

    updateDocument(e) {
        const field = e.target.name;
        const Document = this.state.document;
        document[field] = e.target.value;
        return this.setState({ document: Document });
    }

    onSubmit() {
        this.props.updateDocument(this.state.document);
        this.context.router.push('/documentView');
        this.setState({ open: false });
    }

    handleOpen() {
        this.setState({ open: true });
    }

    handleClose() {
        this.setState({ open: false });
    }

    render() {
        const actions = [
            <FlatButton
                label="Cancel"
                primary
                onTouchTap={this.handleClose}
            />,
            <FlatButton
                label="Submit"
                primary
                keyboardFocused
                onTouchTap={this.handleClose}
            />,
        ];

        return (
            <Dialog
                title="Edit Document"
                actions={actions}
                modal={false}
                open={this.state.open}
                onRequestClose={this.handleClose}
            >
                <div>
                    <EditDocument
                        id={this.props.document.id}
                        title={this.props.document.title}
                        content={this.props.document.content}
                        access={this.props.document.access}
                        onSubmit={this.onSubmit}
                        onChange={this.updateDocument}
                    />
                </div>
            </Dialog>
        );
    }
}

EditDocumentContainer.PropTypes = {
    updateDocument: PropTypes.func.isRequired,
    deleteDocument: PropTypes.func.isRequired,
    listDocuments: PropTypes.func.isRequired,
    actions: PropTypes.object.isRequired,
    document: PropTypes.object.isRequired
};

EditDocumentContainer.contextTypes = {
    router: React.PropTypes.object.isRequired
};

const mapStateToProps = (state, ownProps) => ({
    document: state.document
});

function mapDispatchToProps(dispatch) {
    return {
        actions: bindActionCreators(documentActions, dispatch)
    };
}

export default connect(mapStateToProps, { updateDocument, deleteDocument, listDocuments })(EditDocumentContainer);
