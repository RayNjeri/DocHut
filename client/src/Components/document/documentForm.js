import React, { PropTypes } from 'react';
import TextField from 'material-ui/TextField';
import RaisedButton from 'material-ui/RaisedButton';

const style = {
  margin: 12,
};
class EditDocument extends React.Component {
  render() {
    return (
      <div>
        <form>
          <TextField
            type="text"
            name="title"
            placeholder="Title"
            defaultValue={this.props.title}
            onChange={this.props.onChange}
            fullWidth
          />
          <br />
          <TextField
            type="text"
            name="content"
            placeholder="Content"
            defaultValue={this.props.content}
            onChange={this.props.onChange}
            multiLine
            fullWidth
          />
          <br />
          <TextField
            type="text"
            name="access"
            label="Access"
            defaultValue={this.props.access}
            onChange={this.props.onChange}
            fullWidth
          />
          <br />
          <RaisedButton
            primary
            style={style}
            type="submit"
            label="Update"
            disabled={this.props.saving}
            className="btn btn-primary"
            onClick={this.props.onSubmit}
          />
        </form>
      </div>
    );
  }
}

EditDocument.PropTypes = {
  id: PropTypes.number.isRequired,
  title: PropTypes.string.isRequired,
  content: PropTypes.string.isRequired,
  saving: PropTypes.func.isRequired,
  access: PropTypes.string.isRequired,
  onSubmit: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired
};

export default EditDocument;
