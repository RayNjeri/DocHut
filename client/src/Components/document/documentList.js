import React from 'react';
import PropTypes from 'prop-types';
import FloatingActionButton from 'material-ui/FloatingActionButton';
import ContentAdd from 'material-ui/svg-icons/content/add';
import DocumentView from './documentsView';

const DocumentList = ({ documents, handleOpen, style }) => (
    <div>
        {documents.map(document =>
            <DocumentView key={document.id} document={document} />
        )}
        {handleOpen && (
            <FloatingActionButton onClick={handleOpen} style={style}>
                <ContentAdd />
            </FloatingActionButton>
        )}
    </div>
);
DocumentList.propTypes = {
  documents: PropTypes.array.isRequired,
  handleOpen: PropTypes.func,
  style: PropTypes.func
};

export default DocumentList;
