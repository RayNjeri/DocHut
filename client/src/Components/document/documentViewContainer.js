import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import Pagination from 'react-js-pagination';
import Dialog from 'material-ui/Dialog';
import FlatButton from 'material-ui/FlatButton';
import FloatingActionButton from 'material-ui/FloatingActionButton';
import ContentAdd from 'material-ui/svg-icons/content/add';
import * as documentActions from '../../actions/documentsAction';
import Snackbar from 'material-ui/Snackbar';
import DocumentView from './documentsView';
import CreateDocument from './documentCreateForm';
import DocumentEditForm from './documentEditForm';
import DocumentSearch from './documentSearch';

const style = {
  position: 'fixed',
  bottom: 20,
  right: 20,
  marginRight: 20
};

export class DocumentViewContainer extends React.Component {
  constructor(props, context) {
    super(props, context);
    this.state = {
      open: false,
      activePage: 1,
      limit: 3,
      offset: 0,
      document: {
        title: '',
        content: '',
        access: '',
        snackBarOpen: false
      }
    };
    this.handleOpen = this.handleOpen.bind(this);
    this.handleClose = this.handleClose.bind(this);
    this.onSetAccess = this.onSetAccess.bind(this);
    this.onTitleChange = this.onTitleChange.bind(this);
    this.onContentChange = this.onContentChange.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handlePageChange = this.handlePageChange.bind(this);
    this.closeSnackBar = this.closeSnackBar.bind(this);
  }
  componentWillMount() {
    this.props.documentActions.listDocuments(
      this.state.limit,
      this.state.offset
    );
  }
  onSetAccess(value) {
    const Document = this.state.document;
    Document.access = value;
    this.setState({ document: Document });
  }
  onTitleChange(event) {
    const Document = this.state.document;
    Document.title = event.target.value;
    this.setState({ document: Document });
  }
  onContentChange(event) {
    const Document = this.state.document;
    Document.content = event.target.value;
    this.setState({ document: Document });
  }

  handleOpen() {
    this.setState({ open: true, document: {} });
  }

  handleClose() {
    this.setState({ open: false, edit: false });
  }

  handleChange(event) {
    const { name, value } = event.target;
    this.setState({
      document: Object.assign({}, this.state.document, {
        [name]: value
      })
    });
  }

  updateDocument(document) {
    return () => {
      this.setState({
        document,
        edit: true,
        open: true
      });
    };
  }

  handleSubmit(event) {
    event.preventDefault();
    this.state.edit
      ? this.props.documentActions.updateDocument(this.state.document)
      : this.props.documentActions.createDocument(this.state.document);

    this.setState({ snackBarOpen: true });
    this.handleClose();
  }

  handlePageChange(pageNumber) {
    this.setState({ activePage: pageNumber });
    this.props.documentActions.listDocuments(
      this.state.limit,
      this.state.limit * (pageNumber - 1)
    );
  }

  closeSnackBar() {
    this.setState({ snackBarOpen: false });
  }

  render() {
    const viewActions = [
      <FlatButton label="Cancel" primary onTouchTap={this.handleClose} />,
      <FlatButton
        label="Submit"
        primary
        keyboardFocused
        onTouchTap={this.handleSubmit}
      />
    ];
    return (
      <div className="container">
        <div className="documents">
          <h1>Documents</h1>
          <DocumentSearch />
          {this.props.documentList.documents.map(document => (
            <DocumentView
              key={document.id}
              document={document}
              onUpdate={this.updateDocument(document)}
              deleteDocument={this.props.documentActions.deleteDocument}
              listDocuments={this.props.documentActions.listDocuments}
            />
          ))}
          <div className="paginitaion">
            <Pagination
              activePage={this.state.activePage}
              itemsCountPerPage={this.state.limit}
              totalItemsCount={10}
              pageRangeDisplayed={3}
              onChange={this.handlePageChange}
            />
          </div>
          <div>
            <FloatingActionButton
              onClick={this.handleOpen}
              backgroundColor="#123c69"
              style={style}
            >
              <ContentAdd />
            </FloatingActionButton>
          </div>
        </div>
        <br />
        <Dialog
          title="Create a new Document"
          actions={viewActions}
          modal={false}
          open={this.state.open}
          onRequestClose={this.handleClose}
        >
          {this.state.edit ? (
            <DocumentEditForm
              document={this.state.document}
              onChange={this.handleChange}
            />
          ) : (
            <CreateDocument
              style={style}
              onSetAccess={this.onSetAccess}
              document={this.state.document}
              onTitleChange={this.onTitleChange}
              onContentChange={this.onContentChange}
            />
          )}
        </Dialog>
        <Snackbar
          open={this.state.snackBarOpen}
          message="Document Saved"
          autoHideDuration={2000}
          onRequestClose={this.closeSnackBar}
        />
      </div>
    );
  }
}
DocumentViewContainer.propTypes = {
  documentList: PropTypes.object.isRequired,
  documentActions: PropTypes.object.isRequired
};

function mapStateToProps(state) {
  return {
    documentList: state.documents
  };
}

function mapDispatchToProps(dispatch) {
  return {
    documentActions: bindActionCreators(documentActions, dispatch)
  };
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(DocumentViewContainer);
