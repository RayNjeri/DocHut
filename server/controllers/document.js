const { Document, Role } = require('../models');

// create document

module.exports = {
  create(req, res) {
    const { userId, roleId } = req;
    const data = {
      title: req.body.title,
      content: req.body.content,
      access: req.body.access,
      userId,
      roleId,
    };

    if (req.body.access === 'role') {
      Role.findById(roleId)
        .then((role) => {
          data.access = role.roleName;  
          Document.create(data)
                .then(document => res.status(201).send(document))
                .catch(error => res.status(400).send(error));
        });
    } else {
      Document.create(data)
            .then(document => res.status(201).send(document))
            .catch(error => res.status(400).send(error));  
    }
  },

    // update document by Id

  update(req, res) {
    Document.findById(req.params.documentId, {
    })
    .then((document) => {
      if (!document) {
        return res.status(404).send({
          message: 'Document Not Found',
        });
      }
      document.update(req.body, { fields: Object.keys(req.body)})
      .then(() => res.status(201).send(document))
      .catch(error => {
        console.log("xxxx", error );
        res.status(400).send(error); 
      });
    })
    .catch(error => {
      res.status(400).send(error);
    });
  },

    // list all documents

  list(req, res) {
    const findDocs = (query) => {
      return Document.findAll({
        offset: req.query.offset,
        limit: req.query.limit,
        query})
        .then(response => res.status(200).send(response))
        .catch(error => res.status(400).send(error));
    };
    const querySearch = (offset, limit, isPublic, roleId) => {
      let query = {
        where: {
          $or: []
        }
      };
      if (offset) {
        req.query.offset = offset;
      }
      if (limit) {
        req.query.limit = limit;
      }
      if (isPublic) {
        query.where.$or.push({ access: 'public' });
      }
      if (roleId) {
        Role.findById(roleId)
          .then((role) => {
            query.where.$or.push({ access: role.roleName });
            findDocs(query);
          });
      } else {
        findDocs(query);
      }
    };
    const allSearch = (isPublic) => {
      let query = {};
      if (isPublic) {
        query.where = {
          access: 'public'
        };
      }
      return Document.findAll(query)
                .then(document => res.status(200).send(document))
                .catch(error => res.status(400).send(error));
    };
    if (req.roleId == 1) {
      if (req.query.limit || req.query.offset) {
        querySearch(req.query.offset, req.query.limit);
      } else {
        allSearch();
      }
    } else {
      if (req.query.limit || req.query.offset) {
        querySearch(req.query.offset, req.query.limit, true, req.roleId);
      } else {
        querySearch(null, null, true, req.roleId);
       
      }
    }
  },

    // get a document by id

  retrieve(req, res) {
    Document.findById(req.params.documentId)
            .then((document) => {
              if (!document) {
                return res.status(404).send({
                  message: 'Document Not Found',

                });
              }
              return res.status(200).send(document);
            })
            .catch(error => res.status(400).send(error));
  },

    // delete a document

  destroy(req, res) {
    Document.findById(req.params.documentId)
            .then((document) => {
              if (!document) {
                return res.status(404).send({
                  message: 'Document Not Found',
                });
              }
              document.destroy()
                    .then(() => res.status(204).send({ message: 'Document Succesfully deleted' }))
                    .catch(error => res.status(400).send(error));
            })
            .catch(error => res.status(400).send(error));
  },

    //search document

  findByTitle(req, res) {
    return Document
            .findAll({
              where: {
                title: { $ilike: `%${req.query.q}%` }
              }
            })
            .then(response => res.status(302).send(response))
            .catch(error => res.status(400).send(error));
  },

};

