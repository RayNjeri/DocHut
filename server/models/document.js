module.exports = (sequelize, DataTypes) => {
  const Document = sequelize.define('Document', {
    title: {
      type: DataTypes.STRING,
      allowNull: false
    },
    content: {
      type: DataTypes.TEXT,
      allowNull: false
    },
    access: {
      type: DataTypes.STRING,
      allowNull: false
    },
  }, {
    tableName: 'documents',
    classMethods: {
      associate: (models) => {
        Document.belongsTo(models.User, {
          foreignKey: 'userId',
          onDelete: 'CASCADE',
        });
        Document.belongsTo(models.Role, {
          foreignKey: 'roleId'
        });
      },
    },
  });
  return Document;
};
