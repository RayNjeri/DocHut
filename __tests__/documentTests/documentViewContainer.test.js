import React from 'react';
import { mount, shallow, wrapper } from 'enzyme';
import shallowToJSON from 'enzyme-to-json';
import { DocumentViewContainer } from '../../client/src/Components/document/documentViewContainer';

// function setItem() {
//   const props = {
//     createDocument: jest.fn()
//   };
//   const enzymeWrapper = mount(<DocumentViewContainer {...props}/>);
//   return {
//     props,
//     enzymeWrapper
//   };
// }

// describe('components', () => {
//   describe('DocumentViewContainer', () => {
//     it('should render self and subcomponents', () =>{
//       const { enzymeWrapper } = setItem();
//       expect(enzymeWrapper.find(DocumentViewContainer).hasClass(DocumentViewContainer)).toBe(true);
//     });
//   });
// });

describe('DocumentViewContainer component',() => {
  it('snapshot of the component', () =>{
    const props = {
      documentActions: {
        listDocuments() {}
      },
      documentList: {
        documents: []
      },
    }
    const component = shallow(<DocumentViewContainer {...props} />);
    const tree = shallowToJSON(component);
    expect(tree).toMatchSnapshot();
  });
});
