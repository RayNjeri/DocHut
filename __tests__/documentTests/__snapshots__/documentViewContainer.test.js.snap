// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DocumentViewContainer component snapshot of the component 1`] = `
<div
  className="container"
>
  <div>
    <div>
      <FloatingActionButton
        backgroundColor="#123c69"
        disabled={false}
        mini={false}
        onClick={[Function]}
        secondary={false}
        style={
          Object {
            "bottom": 20,
            "marginRight": 20,
            "position": "fixed",
            "right": 20,
          }
        }
        zDepth={2}
      >
        <ContentAdd />
      </FloatingActionButton>
    </div>
  </div>
  <br />
  <Dialog
    actions={
      Array [
        <FlatButton
          disabled={false}
          fullWidth={false}
          label="Cancel"
          labelPosition="after"
          labelStyle={Object {}}
          onKeyboardFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onTouchStart={[Function]}
          onTouchTap={[Function]}
          primary={true}
          secondary={false}
      />,
        <FlatButton
          disabled={false}
          fullWidth={false}
          keyboardFocused={true}
          label="Submit"
          labelPosition="after"
          labelStyle={Object {}}
          onKeyboardFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onTouchStart={[Function]}
          onTouchTap={[Function]}
          primary={true}
          secondary={false}
      />,
      ]
    }
    autoDetectWindowHeight={true}
    autoScrollBodyContent={false}
    modal={false}
    onRequestClose={[Function]}
    open={false}
    repositionOnUpdate={true}
    title="Create a new Document"
  >
    <CreateDocument
      document={
        Object {
          "access": "",
          "content": "",
          "title": "",
        }
      }
      onContentChange={[Function]}
      onSetAccess={[Function]}
      onTitleChange={[Function]}
      style={
        Object {
          "bottom": 20,
          "marginRight": 20,
          "position": "fixed",
          "right": 20,
        }
      }
    />
  </Dialog>
</div>
`;
